### ISSUE TITLE
.NET MAUI seeing lots of new CA1416 warnings

### ISSUE THREAD

**Issue by jonathanpeppers** [dotnet/roslyn-analyzers#6158]
### Analyzer

**Diagnostic ID**: [CA1416](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1416): Validate platform compatibility

### Analyzer source

**SDK**: Built-in CA analyzers in .NET 7 SDK

**Version**: 7.0.100-rc.2.22457.6

### Describe the bug

We started getting many warnings in dotnet/maui, during this bump: https://github.com/dotnet/maui/pull/9968

Going from .NET 7.0.100-rc.1.22425.9 to 7.0.100-rc.2.22457.6

An example:

```
error CA1416: This call site is reachable on: 'iOS' 10.0 and later, 'maccatalyst' 10.0 and later. 'UIScreen.MaximumFramesPerSecond' is only supported on: 'ios' 10.3 and later, 'maccatalyst' 10.3 and later, 'tvos' 10.2 and later.
```

https://github.com/dotnet/maui/blob/dcd91ec3954b225664be410660ac0d4ade24fc9a/src/Essentials/src/DeviceDisplay/DeviceDisplay.ios.cs#L21-L23

We don't think that the `MaximumFramesPerSecond` API changed any attributes, but the CA1416 appeared.

### Steps To Reproduce

1. Build the branch: https://github.com/dotnet/maui/pull/9968
2. this provisions a .NET SDK & optional workloads, then builds
3. See error

### Expected behavior

I don't think we should be seeing CA1416 several of the cases I've reviewed.

### Actual behavior

We got a lot of CA1416 warnings.

### Other info:

Here is a `.binlog` with several other examples of warnings: [logs.zip](https://github.com/dotnet/roslyn-analyzers/files/9557507/logs.zip)

**Comment by Redth** [comment]
@buyaa-n any updates on this one?

**Comment by buyaa-n** [comment]
> We don't think that the MaximumFramesPerSecond API changed any attributes, but the CA1416 appeared.

I don't think this is caused by the analyzer, 'UIScreen.MaximumFramesPerSecond' must have been annotated somehow.

 I need to build the MAUI repo for finding out what is going on, but had a trouble during build when I tried before, will give it try again later

**Comment by rolfbjarne** [comment]
> > We don't think that the MaximumFramesPerSecond API changed any attributes, but the CA1416 appeared.
> 
> I don't think this is caused by the analyzer, 'UIScreen.MaximumFramesPerSecond' must have been annotated somehow.
> 
> I need to build the MAUI repo for finding out what is going on, but had a trouble during build when I tried before, will give it try again later

You can also download the NuGet to check the attributes: https://www.nuget.org/packages/Microsoft.iOS.Ref/16.0.1478

**Comment by jonathanpeppers** [comment]
@buyaa-n if you look at the usage:

https://github.com/dotnet/maui/blob/63f2ed58ccae3014816da5e03bf5a505a7c40a1a/src/Essentials/src/DeviceDisplay/DeviceDisplay.ios.cs#L21-L23

These all check for 10.3, right? What should we do to avoid the warning in this case?

**Comment by buyaa-n** [comment]
Ah, did not know a conditional involved, I was able to repro, and it looks an analyzer bug.

**Comment by buyaa-n** [comment]
The MAUI repo build became very smooth, this time I was able to build and investigate the warnings.

Two bugs found during investigation for which I [put up a fix](https://github.com/dotnet/roslyn-analyzers/pull/6361)
1. When a child API narrows down parent supported version the guarding version was compared with parent version only when it should be compared with child version. Which was causing the issue: https://github.com/dotnet/maui/blob/dcd91ec3954b225664be410660ac0d4ade24fc9a/src/Essentials/src/DeviceDisplay/DeviceDisplay.ios.cs#L21-L23
    ```cs
    var rate = (OperatingSystem.IsIOSVersionAtLeast(10, 3) || OperatingSystem.IsMacCatalystVersionAtLeast(10, 3) || OperatingSystem.IsTvOSVersionAtLeast(10, 3))
		   ? UIScreen.MainScreen.MaximumFramesPerSecond // warns here when it should not
		   : 0;
    ```
   Note that `OperatingSystem.IsMacCatalystVersionAtLeast(10, 3)` check is redundant here as it is covered by `OperatingSystem.IsIOSVersionAtLeast(10, 3)`

2. When a called API has `SupportedOSPlatform` and `ObsoletedOSPlatform` attributes both and if the callsite attribute suppresses the Supported version the analyzer should warn only for the Obsoleted version. It was still warning for the Supported attribute even if it was empty / suppressed. Example:

    ```cs
    class Program
    {
        [ObsoletedOSPlatform("ios7.0", "Use 'NSString.GetBoundingRect (CGSize, NSStringDrawingOptions, UIStringAttributes, NSStringDrawingContext)' instead.")]
        [ObsoletedOSPlatform("maccatalyst7.0", "Use 'NSString.GetBoundingRect (CGSize, NSStringDrawingOptions, UIStringAttributes, NSStringDrawingContext)' instead.")] // maccatalyst covered by ios, this attribute redundant
        [SupportedOSPlatform("ios")]
        [SupportedOSPlatform("maccatalyst")] // this attribute is redundant, maccatalyst covered by [SupportedOSPlatform("ios")]
        public static void M3() { }
    
        [SupportedOSPlatform("ios")]
        public static void M1()
        {
             M3(); // it was reporting 2 diagnostics here, CA1416 and CA1422 when only CA1422 was expected
        }
    }
     ```

**Comment by buyaa-n** [comment]
The fix is now available on nuget with [8.0.0-preview1.22620.4](https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet8/NuGet/Microsoft.CodeAnalysis.NetAnalyzers/versions/8.0.0-preview1.22627.1) version, in case if needed we might be able to port the fix into .Net7 SDK

**Comment by jonathanpeppers** [comment]
@buyaa-n yes, I think you should probably service this fix in .NET 7. From what I saw it was a regression from .NET 6?

**Comment by buyaa-n** [comment]
I believe bug 1 was in .NET 6, bug 2 was introduced in .NET 7 with `ObsoletedOSPlatform` attribute support, anyway I will ask for servicing
